flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/MultinomialNaiveBayes.scala

```diff
@@ -0,0 +1,900 @@
+          }
+
+          override def map(value: (String, String, Int)): (String, String, Double) = {
+            (value._1, value._2, (value._3 + 1) / (broadcastMap(value._1) + vocabularyCount))
+          }
+        }).withBroadcastSet(allWordsInClass, "allWordsInClass")
+
+      }
+
+      //END POSSIBILITY 3
+
+      //stores all the word related information in one data set
+      // 1. Map: Caluclate logarithms
+      val wordRelatedModelData = pwc.map(line => (line._1, line._2, Math.log(line._3)))
+
+      //store all class related information in one data set
+      // 1. Join: P(c) data set and P(w|c) data set not in class and calculate logarithms
+      val classRelatedModelData = pc.join(pwcNotInClass)
+        .where(0).equalTo(0) {
+        (line1, line2) => (line1._1, Math.log(line1._2), Math.log(line2._2))
+      } // (class name -> log(P(c)) -> log(P(w|c) not in class))
+
+      instance.wordRelatedModelData = Some(wordRelatedModelData)
+      instance.classRelatedModelData = Some(classRelatedModelData)
+
+      //RENNIE 1: transform document frequency
+        //for this, the improvementData set must be set
+        //calculate (word -> log number of documents in all classes / docs with that word)
+
+      val r1 = resultingParameters(R1)
+
+      if (r1 == 1) {
+        val totalDocumentCount: DataSet[(Double)] = documentsPerClass
+          .reduce((line1, line2) => (line1._1, line1._2 + line2._2))
+          .map(line => line._2) // (count of all documents)
+
+        //number of occurences over all documents of all classes
+        val wordCountTotal = input
+          .flatMap(new SingleDistinctWordSplitter())
+          .map(line => (line._2, 1))
+          .groupBy(0)
+          .reduce((line1, line2) => (line1._1, line1._2 + line2._2))
+           // (word -> count of documents with that word)
+
+        val improvementData = wordCountTotal.map(new RichMapFunction[(String, Int),
+          (String, Double)] {
+
+          var broadcastSet: util.List[Double] = null
+
+          override def open(config: Configuration): Unit = {
+            broadcastSet = getRuntimeContext.getBroadcastVariable[Double]("totalDocumentCount")
+            if (broadcastSet.size() != 1) {
+              throw new RuntimeException("The total document count data set used by 11 = 1 has " +
+                "the wrong size! Please use r1 = 0 if the problem can not be solved.")
+            }
+          }
+
+          override def map(value: (String, Int)): (String, Double) = {
+            (value._1, Math.log(broadcastSet.get(0) / value._2))
+          }
+        }).withBroadcastSet(totalDocumentCount, "totalDocumentCount")
+
+        instance.improvementData = Some(improvementData)
+      }
+
+    }
+  }
+
+  // Model (String, String, Double, Double, Double)
+  implicit def predictNNB = new PredictDataSetOperation[MultinomialNaiveBayes,
+    (Int, String), (Int, String)]() {
+
+    override def predictDataSet(instance: MultinomialNaiveBayes,
```

Comment:
Formatting of parameter list: Every parameter should be in a single line if the complete list does not fit in a single line. Take a look at the other Scala code.